packages/muelu/src/Misc/MueLu_FilteredAFactory_def.hpp

```diff
@@ -708,6 +731,215 @@ namespace MueLu {
     GetOStream(Runtime0)<< "Fixing "<< g_fixedDiags<<" zero diagonal values" <<std::endl;
   }
 
+  // fancy lumping trying to not just move everything to the diagonal but to also consider moving
+  // some lumping to the kept off-diagonals. We basically aim to not increase the diagonal
+  // dominance in a row. In particular, the goal is that row i satisfies
+  //
+  //        lumpedDiagDomMeasure_i   <= rho2
+  // or 
+  //        lumpedDiagDomMeasure <= rho*unlumpedDiagDomMeasure
+  //
+  // NOTE: THIS CODE assumes direct access to a row. See comments above concerning
+  //       ASSUME_DIRECT_ACCESS_TO_ROW
+  //
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  void FilteredAFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+  ExperimentalLumping(const Matrix& A, Matrix& filteredA, double irho, double irho2) const {
+    using TST = typename Teuchos::ScalarTraits<SC>;
+    SC zero = TST::zero();
+    SC one  = TST::one();
+
+    ArrayView<const LO> inds;
+    ArrayView<const SC> vals;
+    ArrayView<const LO> finds;
+    ArrayView<SC> fvals;
+
+    SC PosOffSum, NegOffSum, PosOffDropSum,  NegOffDropSum;
+    SC diag, gamma, alpha; 
+    LO NumPosKept, NumNegKept;
+
+    SC noLumpDdom;
+    SC numer,denom;
+    SC PosFilteredSum, NegFilteredSum;
+    SC Target;
+
+    SC rho  = as<Scalar>(irho);
+    SC rho2 = as<Scalar>(irho2);
+
+    for (LO row = 0; row < (LO) A.getRowMap()->getNodeNumElements(); row++) {
+        noLumpDdom = as<Scalar>(10000.0);  // only used if diagonal is zero
```

Comment:
I've added the following comment in the code near the 10K assignment

the whole idea sort of breaks down when the matrix diagonal is zero. In particular, the old diag dominance ratio is infinity ... so what do we want for the new ddom ratio? Do we want to allow the diagonal to go negative, just to have a better ddom ratio? The current choice essentially changes 'Target' to a large number meaning that we will allow the new ddom number to be fairly large (because the old one was infinity)